### PR TITLE
[Dynamic Region]: bucket key range should be updated after ingested sst

### LINKED ISSUE #15147: [Dynamic Regions] Let TiKV update buckets ASAP after ingesting SST files
## Bug Report

BR restore meets checksum timeout frequently because of a large bucket covers
too many key-value pairs, this indicates that buckers are not updated in time after 
ingesting SST files.

Also lightning meets analyze timeout after importing data, they are caused by 
the same reason.

### What version of TiKV are you using?

v7.2.0

### Steps to reproduce

BR restore a large table.

### What did you expect?

No checksum error.

### What did happened?

Checksum error.

### PR DESCRIPTION
**Pull request by bufferflies**
### What is changed and how it works?

Issue Number: Close #15147, #15106

What's Changed:

```commit-message
bucket write keys and bytes should be updated after ingested sst. So  the bucket key range will be refreshed .
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

master:
img_v2_21fc787e-7c8e-4b8c-b75b-87d1baab142g
img_v2_d65e0900-8eec-490b-896a-4ef45d34823g

this pr: 
img_v2_22f00d72-0641-495f-a1ef-1f08b50a8c4g
img_v2_21fc787e-7c8e-4b8c-b75b-87d1baab142g

Side effects

### Release note 

```release-note
bucket write key and bytes should be recorded after sst ingested.
```

### CODE REVIEW

File: tests/failpoints/cases/test_import_service.rs [hunk 1]
```diff
@@ -397,6 +371,55 @@ fn test_delete_sst_after_applied_sst() {
...
+        tx.lock().unwrap().send(()).unwrap();
+    })
+    .unwrap();
+    rx.recv_timeout(std::time::Duration::from_millis(100))
+        .unwrap();
+    std::thread::sleep(std::time::Duration::from_millis(100));
+    let region_keys = cluster
+        .pd_client
+        .get_region_approximate_keys(ctx.get_region_id())
+        .unwrap();
```
Comment (overvenus) [inline tests/failpoints/cases/test_import_service.rs:416]: Might be unstable, could you retry the check in a loop and panic it fails the check?

File: tests/integrations/import/test_sst_service.rs [hunk 2]
```diff
@@ -118,6 +127,14 @@ fn test_ingest_sst() {
     ingest.set_sst(meta);
     let resp = import.ingest(&ingest).unwrap();
     assert!(!resp.has_error(), "{:?}", resp.get_error());
+    std::thread::sleep(std::time::Duration::from_millis(100));
+    let region_keys = cluster
+        .pd_client
+        .get_region_approximate_keys(ctx.get_region_id())
+        .unwrap();
+    assert_eq!(255, region_keys);
```
Comment (overvenus) [inline tests/integrations/import/test_sst_service.rs:135]: doitto

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: tests/failpoints/cases/test_import_service.rs (hunk 1 above)
Comment (bufferflies) [reply tests/failpoints/cases/test_import_service.rs:416]: fixed

File: tests/integrations/import/test_sst_service.rs (hunk 2 above)
Comment (bufferflies) [reply tests/integrations/import/test_sst_service.rs:135]: fixed

File: components/pd_client/src/lib.rs [hunk 3]
```diff
@@ -198,6 +198,15 @@ impl BucketStat {
         }
     }
 
+    pub fn ingest_sst(&mut self, key: u64, value_size: u64) {
```
Comment (tonyxuqqi) [inline components/pd_client/src/lib.rs:201]: key may be renamed as key_size

Comment (tonyxuqqi) [reply components/pd_client/src/lib.rs:201]: Actually from the caller code the 'key' parameter actually is the keys count. Then the code below is incorrect---it should not be write_bytes.
 for stat in self.stats.mut_write_bytes() {
            *stat += key;
        }

Comment (bufferflies) [reply components/pd_client/src/lib.rs]: good catch, fixed

File: components/pd_client/src/lib.rs [hunk 4]
```diff
@@ -198,6 +198,15 @@ impl BucketStat {
         }
     }
 
+    pub fn ingest_sst(&mut self, key_count: u64, value_size: u64) {
+        for stat in self.stats.mut_write_bytes() {
+            *stat += value_size;
+        }
```
Comment (overvenus) [inline components/pd_client/src/lib.rs:204]:
```suggestion
        let first_idx = find_bucket_index(sst.first_key, &self.meta.keys);
        let last_idx = find_bucket_index(sst.last_key, &self.meta.keys);
        if last_idx - first_idx == 0 {
            if let Some(bytes) = self.stats.mut_write_bytes().get_mut(first_idx) {
                *bytes += value_size;
            }
        } else { 
            let avg = value_size/(last_idx - first_idx);
            for idx in first_idx..=last_idx {
                if let Some(bytes) = self.stats.mut_write_bytes().get_mut(idx) {
                    *bytes += avg;
                }
            }
        }
```

Maybe we should only add size to buckets that covers ingested ssts.

Comment (bufferflies) [reply components/pd_client/src/lib.rs:206]: no, the condition of data distribution in all the buckets is not true.

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**